### PR TITLE
feat: add rhoas openapi validator

### PR DESCRIPTION
**Pull request by craicoverflow**
You can play with this using `npx`: 

```shell
npx @rhoas/openapi-validator@0.0.2-dev1 validate openapi.yaml
```

## Rules

The RHOAS ruleset extends the Spectral built-in "oas" ruleset (except `operation-tags`, `openapi-tags`). You can see the full list of rules from that ruleset [here](https://meta.stoplight.io/docs/spectral/docs/reference/openapi-rules.md)

### oas3minimum

OpenAPI schemas should be a minimum of v3.

```yaml
openapi: 3.0
```

**Recommended**: Yes
**Severity**: warning

### servers-config

The `servers` OpenAPI object must be defined and must specify at minimum the following URLs:

```yaml
servers:
- url: https://api.openshift.com
- url: https://api.stage.openshift.com
- url: http://localhost:8000
- url: /
```

**Recommended**: Yes
**Severity**: warning

### info-license-apache2.0:

The `info.license.name` field must be "Apache 2.0".

```yaml
info:
  license:
    name: 'Apache 2.0'
```

**Recommended**: Yes
**Severity**: warning

### info-license-apache2.0-url:

The `info.license.url` field must have the correct link for Apache 2.0.

```yaml
info:
  license:
    url: 'https://www.apache.org/licenses/LICENSE-2.0.html'
```

**Recommended**: Yes
**Severity**: warning

### invalid-path-regexp

All paths must match the specified regular expression: `/api/([a-z_]*){1,}(/v[0-9]*(alpha|beta)?)(/{?[a-z_]*}?){1,}$"`.

- The first segment must be `/api`
- The second segment can only contain alphabetical characters and underscores "_"
- The third segment must specify the API version. This can be a major version such as `v1` or a channel-version such as `v1beta`, `v1alpha`.
- All following segments must follow `camel_case` and can only contain alphabetical characters.

**Recommended**: Yes
**Severity**: warning

### invalid-response-media-type

The content type for all responses must be `application/json`.

**Recommended**: Yes
**Severity**: error

### invalid-error-response-object

All error response bodies must reference `#/components/Schemas/Error`

**Recommended**: Yes
**Severity**: error

### invalid-object-resource-schema

All API response bodies must be an `object` with three required properties:

```yaml
type: object
required: [id, kind, href]
properties:
  id:
    type: string
  kind:
    type: string
  href:
    type: string
```

**Recommended**: Yes
**Severity**: error

### schema-name-camel-case

All JSON schema objects defined in `components.schemas` must follow `CamelCase`.

**Recommended**: Yes
**Severity**: warning
### properties-snake-case

All JSON schema properties defined must follow `camel_case`.

**Recommended**: Yes
**Severity**: error

### invalid-error-schema

`components.schema` MUST have a valid `Error` object.

```yaml
Error:
  type: object
  required: [id, kind, href, code, reason]
  properties:
    id:
      type: string
    kind:
      type: string
    href:
      type: string
    code:
      type: string
    reason:
      type: string
```

**Recommended**: Yes
**Severity**: warning

### invalid-object-schema

`components.schema` MUST have a valid `ObjectReference` object.

```yaml
ObjectReference:
  type: object
  required: [id, kind, href]
  properties:
    id:
      type: string
    kind:
      type: string
    href:
      type: string
```

**Recommended**: Yes
**Severity**: warning

### invalid-list-schema

`components.schema` MUST have a valid `List` object.

```yaml
List:
  required:
    - kind
    - page
    - size
    - total
    - items
  type: object
  properties:
    items:
      type: array
    kind:
      type: string
    page:
      type: integer
    size:
      type: integer
    total:
      type: integer
```

**Recommended**: Yes
**Severity**: warning

### CODE REVIEW

**Comment by craicoverflow** [comment]
I just noticed that this will not work outside of this repo due to missing dependencies:

```yaml
❯ spectral lint .openapi/*
Cannot find module 'json-schema-resolve-allof'
Require stack:
- /home/ephelan/.nvm/versions/node/v14.15.2/lib/node_modules/@stoplight/spectral/dist/rulesets/evaluators.js
- /home/ephelan/.nvm/versions/node/v14.15.2/lib/node_modules/@stoplight/spectral/dist/spectral.js
- /home/ephelan/.nvm/versions/node/v14.15.2/lib/node_modules/@stoplight/spectral/dist/cli/services/linter/linter.js
- /home/ephelan/.nvm/versions/node/v14.15.2/lib/node_modules/@stoplight/spectral/dist/cli/services/linter/index.js
- /home/ephelan/.nvm/versions/node/v14.15.2/lib/node_modules/@stoplight/spectral/dist/cli/commands/lint.js
- /home/ephelan/.nvm/versions/node/v14.15.2/lib/node_modules/@stoplight/spectral/dist/cli/index.js
```

I think that dependency is required..I guess maybe I can try to add `node_modules` here: https://unpkg.com/browse/@rhoas/spectral-ruleset@0.0.1/

File: spectral/README.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    type: string
+  href:
+    type: string
+  code:
+    type: string
+  reason:
+    type: string
+```
+
+> NOTE: For 5xx errors the `reason` property is not required.
```
Comment (wtrocki) [inline spectral/README.md:139]: That is specific implememtation detail.
None of the fields are marked as required.

**Comment by wtrocki** [comment]
Amazing Documentation and set of the rules. I do not have any specific comments about those.

Question? I'm getting number of issues in the OpenAPI schema that are not listed as rules we enforce. 
Do we also have some config for generic OpenAPI errors etc. 

```
[wtrocki@graphapi app-services-sdk-js (main)]$ npx @rhoas/openapi-validator@0.0.2-dev1 validate .openapi/kas-fleet-manager.yaml 
/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  2:5  warning  OpenAPI `license` object is required   info

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  16:26  warning        OpenAPI paths must start with `/api/`, followed by a version. All paths must follow camel_case   paths./api/kafkas_mgmt/v1

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  34:35  error  missing required property "id"   paths./api/kafkas_mgmt/v1/status.get.responses.200.content.application/json.schema.properties

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  274:275  error        missing required property "id"   paths./api/kafkas_mgmt/v1/kafkas.get.responses.200.content.application/json.schema

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  328:329  error        missing required property "id"   paths./api/kafkas_mgmt/v1/cloud_providers.get.responses.200.content.application/json.schema

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  362:363  error        missing required property "id"   paths./api/kafkas_mgmt/v1/cloud_providers/{id}/regions.get.responses.200.content.application/json.schema

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  392:393  error        missing required property "id"   paths./api/kafkas_mgmt/v1/service_accounts.get.responses.200.content.application/json.schema

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  606:607  error        missing required property "href"   paths./api/kafkas_mgmt/v1/kafkas/{id}/metrics/query_range.get.responses.200.content.application/json.schema

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  642:643  error        missing required property "href"   paths./api/kafkas_mgmt/v1/kafkas/{id}/metrics/query.get.responses.200.content.application/json.schema

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  686:694  error        missing required property "items"   components.schemas.List.properties

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  1004:1004  error      `Timestamp` MUST follow snake_case and be ASCII alphanumeric characters.   components.schemas.MetricsRangeQueryList.allOf.0.properties.items.items.allOf.0.properties.values.items.properties.Timestamp

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  1004:1004  error      `Value` MUST follow snake_case and be ASCII alphanumeric characters.   components.schemas.MetricsRangeQueryList.allOf.0.properties.items.items.allOf.0.properties.values.items.properties.Value

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  1014:1015  error      `Timestamp` MUST follow snake_case and be ASCII alphanumeric characters.   components.schemas.RangeQuery.properties.values.items.properties.Timestamp

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  1014:1015  error      `Value` MUST follow snake_case and be ASCII alphanumeric characters.   components.schemas.RangeQuery.properties.values.items.properties.Value

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  1016:1034  error      `values` object name must follow CamelCase   components.schemas.values

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  1019:1022  error      `Timestamp` MUST follow snake_case and be ASCII alphanumeric characters.   components.schemas.values.properties.Timestamp

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  1026:1029  error      `Value` MUST follow snake_case and be ASCII alphanumeric characters.   components.schemas.values.properties.Value

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  1052:1052  error      `Timestamp` MUST follow snake_case and be ASCII alphanumeric characters.   components.schemas.MetricsInstantQueryList.allOf.0.properties.items.items.allOf.0.properties.Timestamp

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  1052:1052  error      `Value` MUST follow snake_case and be ASCII alphanumeric characters.   components.schemas.MetricsInstantQueryList.allOf.0.properties.items.items.allOf.0.properties.Value

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  1060:1063  error      `Timestamp` MUST follow snake_case and be ASCII alphanumeric characters.   components.schemas.InstantQuery.properties.Timestamp

/Users/wtrocki/Projects/mas/sdks/app-services-sdk-js/.openapi/kas-fleet-manager.yaml
  1067:1070  error      `Value` MUST follow snake_case and be ASCII alphanumeric characters.   components.schemas.InstantQuery.properties.Value

✖ 21 problems (19 errors, 2 warnings)

```

File: openapi-validator/README.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
+# RHOAS OpenAPI Validator
+
+A CLI for validating OpenAPI specifications against the RHOAS API Guidelines.
+
+## Development
```
Comment (wtrocki) [inline openapi-validator/README.md:5]: Reorg - Using should be first dev chapter later.

File: openapi-validator/README.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+## Using
+
+It is recommended to use `npx` to validate your documents to ensure you use the latest validation rules:
+
+```shell
+npx @rhoas/openapi-validator validate ./path/to/openapi.yaml
+```
+
+## Rules
```
Comment (wtrocki) [inline openapi-validator/README.md:37]: [Later] Worth blog posting considering getting approval from some parties.

File: openapi-validator/README.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+The RHOAS ruleset extends the Spectral built-in "oas" ruleset (except `operation-tags`, `openapi-tags`). You can see the full list of rules from that ruleset [here](https://meta.stoplight.io/docs/spectral/docs/reference/openapi-rules.md)
+
+### oas3minimum
```
Comment (wtrocki) [inline openapi-validator/README.md:41]: Do you think it is worth to have fast/small docusaurus webpage that doesn't have front (always goes to the docs).

File: openapi-validator/src/spectral/functions/infoLicenseApache2.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import { IFunctionPaths, IFunctionResult } from "@stoplight/spectral";
+
+export default (targetVal: any, opts: any, paths: IFunctionPaths): IFunctionResult[] => {
```
Comment (wtrocki) [inline openapi-validator/src/spectral/functions/infoLicenseApache2.ts:3]: Would be cool to add comment linking to readme chapter or website in future.

**Comment by wtrocki** [comment]
Seriously good job! Amount of work that was put into this is truly impressive. I left some inspiration contents. 

I have checked that out locally and mingled with valid schema making it invalid etc.

**Comment by craicoverflow** [comment]
> Question? I'm getting number of issues in the OpenAPI schema that are not listed as rules we enforce.
> Do we also have some config for generic OpenAPI errors etc.

Those are all errors from the RHOAS ruleset. I've created a new dev version which also prints the rule ID: `npx @rhoas/openapi-validator@0.0.2-dev2`.

> Is there way to disable those problems for now and create issue to address some of them we want to fix?

A challenge is that we have `error` severity rules but existing APIs may be incompatible and cannot change so it should probably be a warning for them. I could not see a way to override this unless using Spectral directly.

Another option is to not allow the `Error` schema to use `allOf` and continue to use Spectral - this would allow for no external dependencies.

**Pull request merged**